types/src/byzantine.rs

```diff
@@ -109,26 +109,78 @@ impl FraudProof for BadEncodingFraudProof {
         let root = NamespacedHash::try_from(cid.hash().digest())?;
 
         // verify that Merkle proofs correspond to particular shares.
-        for share in &self.shares {
-            share
-                .proof
-                .verify_range(
-                    &root,
-                    &[share.leaf.share.as_ref()],
-                    share.leaf.namespace.into(),
-                )
+        for maybe_share in &self.shares {
+            let Some(share_with_proof) = maybe_share else {
+                continue;
+            };
+            let ShareWithProof {
+                leaf: NmtLeaf { namespace, share },
+                proof,
+            } = share_with_proof;
+            proof
+                .verify_range(&root, &[&share], **namespace)
                 .map_err(Error::RangeProofError)?;
         }
 
-        // TODO: Add leopard reed solomon decoding and encoding of shares
-        //       and verify the nmt roots.
+        // rebuild the whole axis
+        let mut rebuilt_shares: Vec<_> = self
+            .shares
+            .iter()
+            .map(|maybe_share| {
+                maybe_share
+                    .clone()
+                    .map(|share_with_proof| share_with_proof.leaf.share)
+                    .unwrap_or_default()
+            })
+            .collect();
+        let ods_width = rebuilt_shares.len() / 2;
+        if leopard_codec::reconstruct(&mut rebuilt_shares, ods_width).is_err() {
+            // we couldn't reconstruct the data even tho we had enough *proven* shares
+            // befp is legit
+            return Ok(());
```

Comment:
So is this `Ok(())` because we confirm that there is a bad encoding?

Reply:
yes